src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java

```diff
@@ -456,20 +465,109 @@ public void run() {
     }
 
     // All native libraries we've loaded.
-    // This also serves as the lock to obtain nativeLibraries
-    // and write to nativeLibraryContext.
-    private static final Set<String> loadedLibraryNames = new HashSet<>();
+    private static final Set<String> loadedLibraryNames =
+            ConcurrentHashMap.newKeySet();
+
+    // reentrant lock class that allows exact counting (with external synchronization)
+    @SuppressWarnings("serial")
+    private static final class CountedLock extends ReentrantLock {
+
+        private int counter = 0;
+
+        public void increment() {
+            if (++counter < 0) // overflow
+                throw new Error("Maximum lock count exceeded");
+        }
+
+        public void decrement() {
+            if (--counter < 0) // underflow
+                throw new Error("Lock count is below zero");
+        }
+
+        public int getCounter() {
+            return counter;
+        }
+    }
+
+    // Maps native library name to the corresponding lock object
+    private static final Map<String, CountedLock> nativeLibraryLockMap =
+            new ConcurrentHashMap<>();
+
+    private static void acquireNativeLibraryLock(String libraryName) {
+        nativeLibraryLockMap.compute(libraryName, (name, lock) -> {
```

Comment:
This would be clearer if lock were named currentLock as for releaseNLLock

Reply:
David, lock is now renamed to currentLock